File: 4주차/박민형/다리를 지나는 트럭.py
# 1. bridge 변수 사용
# 2. bridge_length 만큼 bridge에 트럭이 있다면 bridge shift
# 3. 새로운 트럭을 넣어줄 때 다리 위의 트럭의 무게 총합과 현재 트럭의 무게를 더했을 때
# => 더한 값이 다리가 버틸 수 있는 무게를 초과한다면 pass
# => 초과하지 않는다면 트럭 넣어주기

from collections import deque

def solution(bridge_length, weight, truck_weights):
    bridge = deque([])
    truck_weights_queue = deque(truck_weights)
    bridge_weight = 0
    answer = 0

    while len(truck_weights_queue) > 0:
        answer += 1
        if len(bridge) == bridge_length:
            shift_weight = bridge.popleft()
            bridge_weight -= shift_weight

        if bridge_weight + truck_weights_queue[0] > weight:
            bridge.append(0)
            continue

        truck_weight = truck_weights_queue.popleft()
        bridge.append(truck_weight)
        bridge_weight += truck_weight

    return answer + bridge_length

File: [2023.09.11 - 2023.09.17] - 2주차/부끄러운어피치/다리를지나는트럭.java
import java.util.LinkedList;
import java.util.Queue;

public class 다리를지나는트럭 {
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        //다리길이, 견딜수있는 무게, 트럭무게배열
        int answer = 0;

        Queue<currentTruck> bridge = new LinkedList<>();
        int currentWeight = 0;
        int passedTruck = 0 ;
        int currentTime = 0;
        //모든 트럭이 건널때까지.... 제한을 주기위함
        while (passedTruck < truck_weights.length) {
            currentTime++;//경과시간 한 반복당 +1
            //트럭이 다리에서 빠져나갈때
            //1. 현재 시간과 큐에서 빠져나가려는 트럭의 진입 시간의 차가 bridge_length와 동일해야 한다.
            if (!bridge.isEmpty() && currentTime - bridge.peek().time == bridge_length) {
                currentTruck truck = bridge.poll();
                currentWeight -= truck.weight;//현재 다리에 가해진무게 - 빠져나간 트럭의 무게
            }

            //지나고있는 사이즈가 최대사이즈보다크면 다음턴
            if (bridge.size() >= bridge_length) {
                continue;
            }

            //현재 다리에 가해진 무개에 다음 트럭이 올라탔을때 > 다리가 견딜수있는 최대 무게면 다음턴
            if (currentWeight + truck_weights[passedTruck] > weight) {
                continue;
            }
            //현재 다리에 가해진무게에 추가 트럭무게를 더한다.
            currentWeight += truck_weights[passedTruck];
            //다리에 진입한 트럭의 진입시간
            //현재 다리에 가해진 무게와, 사이즈, 진입시간 초기화를 위한 로직
            currentTruck currentTruck = new currentTruck();
            currentTruck.setWeight(truck_weights[passedTruck]);
            currentTruck.setTime(currentTime);
            bridge.add(currentTruck);

            passedTruck++;
        }
        //현재 시간 + 마지막 트럭 시간
        return currentTime + bridge_length;
    }

    class currentTruck {
        private int weight;
        private int time;

        public int getWeight() {
            return this.weight;
        }

        public void setWeight(int weight) {
            this.weight = weight;
        }

        public int getTime() {
            return this.time;
        }

        public void setTime(int time) {
            this.time = time;
        }

    }
}